pkg/base/k8s/kubeclient.go

```diff
@@ -110,42 +115,86 @@ func (k *KubeClient) CreateCR(ctx context.Context, name string, obj k8sCl.Object
 	})
 	crKind := obj.GetObjectKind().GroupVersionKind().Kind
 	ll.Infof("Creating CR '%s': %s", crKind, k.objectsLogger.Log(obj))
-	err := k.Create(ctx, obj)
-	if err != nil {
-		if k8sError.IsAlreadyExists(err) {
-			ll.Infof("CR %s %s already exist", crKind, name)
-			return nil
+
+	opt := mergeKubeClientRequestOptions(opts...)
+	var retries int
+	for {
+		err := k.Create(ctx, obj)
+		if err != nil {
+			if k8sError.IsAlreadyExists(err) {
+				ll.Infof("CR %s %s already exist", crKind, name)
+				return nil
+			}
+
+			ll.WithFields(logrus.Fields{"method": "CreateCR", "requestUUID": requestUUID.(string)}).
+				Errorf("error occurred while creating custom resource: '%v'", err)
+			if backoffErr := k.makeBackoffRetry(ctx, "CreateCR", requestUUID.(string), retries, *opt.MaxBackoffRetries); backoffErr != nil {
+				return errors.Wrap(backoffErr, fmt.Sprintf("backoff retries failed for origin error: '%v'", err.Error()))
+			}
 		}
-		ll.Errorf("Unable to create CR %s %s: %v", crKind, name, err)
-		return err
+		ll.Infof("CR %s %s created", crKind, name)
 	}
-	ll.Infof("CR %s %s created", crKind, name)
-	return nil
 }
 
 // ReadCR reads specified resource from k8s cluster into a pointer of struct that implements runtime.Object
 // Receives golang context, name of the read object, and object pointer where to read
 // Returns error if something went wrong
-func (k *KubeClient) ReadCR(ctx context.Context, name string, namespace string, obj k8sCl.Object) error {
+func (k *KubeClient) ReadCR(ctx context.Context, name string, namespace string, obj k8sCl.Object, opts ...*KubeClientRequestOptions) error {
 	defer k.metrics.EvaluateDurationForMethod("ReadCR")()
+	requestUUID := ctx.Value(base.RequestUUID)
+	if requestUUID == nil {
+		requestUUID = DefaultVolumeID
+	}
 	if namespace == "" {
-		return k.Get(ctx, k8sCl.ObjectKey{Name: name, Namespace: k.Namespace}, obj)
+		namespace = k.Namespace
+	}
+
+	opt := mergeKubeClientRequestOptions(opts...)
+	var retries int
+	for {
+		err := k.Get(ctx, k8sCl.ObjectKey{Name: name, Namespace: namespace}, obj)
+		if err == nil {
+			return nil
+		}
+
+		k.log.WithFields(logrus.Fields{"method": "ReadCR", "requestUUID": requestUUID.(string)}).
+			Errorf("error occurred while reading custom resource: '%v'", err)
+		if backoffErr := k.makeBackoffRetry(ctx, "ReadCR", requestUUID.(string), retries, *opt.MaxBackoffRetries); backoffErr != nil {
+			return errors.Wrap(backoffErr, fmt.Sprintf("backoff retries failed for origin error: '%v'", err.Error()))
```

Comment:
I think `%w` error verbs should be used here.

Reply:
we have two errors here, %w is suitable only for 1 error cases